### PR TITLE
New Functionality: Background Variant Resolver

### LINKED ISSUE #93: Background Variant - Alternate?
Is there a graceful way to alternate the background variant? It would have been nice if it accepted a function (like the "serializer" property) but I imagine that would be a breaking change.

### PR DESCRIPTION
**Pull request by mavickers**
### Provide an option to have the background class for each list item be determined by use of an injected function.

**Use-case**: ability to have different color list items.

**Summary**:
- Resolves #93 
- Added backgroundVariantResolver prop in top-level component which passes straight through to the list-item component.
- If both backgroundVariant and backgroundVariantResolver values are specified, the one from the resolver will be favored. Non-string and empty string values from the resolver are ignored.
- The array of base classes in the list item component have been pulled out of the textClasses() method and put into component data. This made it easier to write the tests and may make it easier to inject those classes at a later time.
- Added a series of tests covering a handful of scenarios.
- Updated reference documentation.

**Notes**
- I wanted to add a working example but I could not get vuepress working on my Mac.
- "backgroundVariantResolver" is clunky - it won't hurt my feelings if you want the name changed.
- I didn't make any change to version number - I'll leave that to you.

### CODE REVIEW

File: src/components/VueTypeaheadBootstrapListItem.vue [hunk 1]
```diff
@@ -40,15 +40,29 @@ export default {
...
+    backgroundVariantResolver: {
+      type: Function,
+      default: (d) => null,
+      validator: d => d instanceof Function
+    },
     textVariant: {
       type: String
     }
   },
+  data: function() {
```
Comment (mavickers) [inline src/components/VueTypeaheadBootstrapListItem.vue:52]: I pulled these static classes out of textClasses() and put them into data - makes it easier to possibly inject these values in the future, and makes it easier to run some test scenarios.

**Comment by mattzollinhofer** [comment]
Thanks for this really thorough looking PR! I haven't had a chance to look at it much. 

My initial impression is that I hesitate to add functionality for your use case because it isn't supported (as far as I can tell) in bootstrap's list itself. The goal of this project is to add a typeahead capability, not necessarily to extend the styling capabilities. 

I'm looking to their docs (https://getbootstrap.com/docs/5.0/components/list-group/) to see what options there are for display. 

What do you think?

**Comment by mavickers** [comment]
Hi Matt. What I'm doing here is allowing you to inject a function to determine the classing/coloring of each item in the list, so if you wanted rainbow colors in the typeahead you could do that. For instance, in the list-group link you just referenced you'd be able to color the list options like they are colored in the "contextual classes" section.

There is a "backgroundVariant" prop already in the component that passes a static class name to be used for the list items. In essence this new functionality allows you to pass a function instead which will return the class name.

My specific use case is that I wanted to show certain items in the typeahead list as already in use or occupied. The items are white by default, but if they are in an "occupied" state then they are colored gray (specifically the bootstrap "light" class).

**Comment by mattzollinhofer** [comment]
Yeah, thanks for the followup thoughts. I realized that I totally mis-read some of the docs that I linked to. I didn't think the `list-group` supported colors for individual list items. I think what you've got here is really good. I'm going to merge it in after I update an example to show this usage.

Honestly, looking more closely at this makes me think this component should be using `list-group-item-${context}` rather than `bg-${context}`, but I'm not quite ready to make that change as I'm not 100% what the impact of that would be.

Thanks!

**Comment by mattzollinhofer** [comment]
Actually, when updating the example to use `light` vs `dark` I think we need to use `list-group-item-${context}` because when you only specify `bg-dark` the color of the text isn't adjusted to match and it's unreadable.

**Comment by mattzollinhofer** [comment]
Alright, merged in a pushed a [new version to npm](https://www.npmjs.com/package/vue-typeahead-bootstrap), give it a shot. You can also see the usage of it in the [API Example in the docs](https://mattzollinhofer.github.io/vue-typeahead-bootstrap-docs/examples/examples.html#api-example).

Closes #93

Thanks again for the great work!

**Comment by mavickers** [comment]
> Honestly, looking more closely at this makes me think this component should be using `list-group-item-${context}` rather than `bg-${context}`, but I'm not quite ready to make that change as I'm not 100% what the impact of that would be.
> 

Thanks for the merge! 

I used bg-${context} because that is what the backgroundVariant prop is using. If you change backgroundVariantResolver to use list-group-item-${context} instead you'll probably want to apply that change to backgroundVariant as well.

I noticed another snafu when using I think 'white', that being the hover state didn't look any different. I have a feeling there are a couple bugs in the original css.

**Comment by mattzollinhofer** [comment]
Yeah, you were just following the convention of this component. I just hadn't noticed it until now! It's kind of weird taking over someone else's work, there's just stuff you don't see along the way.

Interesting about `white`. With this change, that just won't work anymore because `white` isn't an option in the `list-group` [options](https://getbootstrap.com/docs/4.1/components/list-group/#contextual-classes). But that reminds me that I need the documentation to no longer point people to the `bg` styles.

**Pull request closed without merging**